### PR TITLE
Fix lack of type resolution issue due to lack of imports. #275

### PR DESCRIPTION
**Pull request by gayanper**
## What it does
The changes will add import statements for found types while traversing symbols in the binary class definition so that the evaluation snippet will properly compiled without needing to have FQN names.

## How to test
Unit tests and sample project

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by gayanper** [comment]
@testforstephen will you have time to have a look at this ?

**Comment by testforstephen** [comment]
Hi @gayanper,

The approach you use to dynamically add the import statements for the evaluation snippets looks good. I think we can apply it to the case in the issue #186 as well, where we can resolve the unresolved types and automatically append their fully qualified names in the import statements.

However, I have some questions about the use cases you used in the Java8Tests.java file. Both of the snippets `new Observation((Object) subject, (Consumer) action)` and `((Subject)this.subject).getName()` work fine for me with or without this PR.

Could you provide more use cases that demonstrate the problem you want to fix in this PR? Thank you.

**Comment by testforstephen** [comment]
>  Both of the snippets new Observation((Object) subject, (Consumer) action) and ((Subject)this.subject).getName() work fine for me with or without this PR.

It works because I attached `debug-lib-sources.jar` to `debug-lib.jar`, where the eval will use SourceBasedSourceGenerator and works fine. Removing the source attachment can reproduce the issue you mentioned.

However, compared to SourceBasedSourceGenerator, the BinaryBasedSourceGenerator still has some limitation. For example, `new Observation((Object) subject, (Consumer) action)` snippet works fine in the Line 19 breakpoint of Observation for source attached sample, but still not work for binary sample.

**Comment by gayanper** [comment]
> 
> However, compared to SourceBasedSourceGenerator, the BinaryBasedSourceGenerator still has some limitation. For example, `new Observation((Object) subject, (Consumer) action)` snippet works fine in the Line 19 breakpoint of Observation for source attached sample, but still not work for binary sample. 

You mean the snippet already in the unit test or something different?

**Comment by testforstephen** [comment]
I'm using the new unit test added your PR.

**Comment by gayanper** [comment]
@testforstephen do you mean the unit test fail when you execute or it fails when you try run the changes in workspace and try it on a real project debugging scenario?

**Comment by testforstephen** [comment]
> @testforstephen do you mean the unit test fail when you execute or it fails when you try run the changes in workspace and try it on a real project debugging scenario?

The existing test cases you added can pass. My point is that there are other cases not covered well by this PR.

Below is the eval comparison between SourceBasedSourceGenerator and BinaryBasedSourceGenerator.
- The eval snippet `new Observation((Object) subject, (Consumer) action)` works fine if the library source jar exists. SourceBasedSourceGenerator is used when the library source is available.
image

- The eval snippet `new Observation((Object) subject, (Consumer) action)` fails if the library source is unavailable. BinaryBasedSourceGenerator is used if the source is unavailable.
image

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/eval/Java8Tests.java [hunk 1]
```diff
@@ -241,4 +241,38 @@ public void testContextEvaluations() throws Exception {
...
 	}
 
+	public void testEvaluate_GH275_StaticBinaryMethod_EvaluateSnippetWithImportedTypes() throws Exception {
+		IJavaThread javaThread = null;
+		try {
+			IJavaLineBreakpoint bp = createLineBreakpoint(getType("com.debug.test.Observation"), 25);
+			javaThread = launchToLineBreakpoint("GH275", bp);
+			assertNotNull("The program did not suspend", javaThread);
+
+			String snippet = "new Observation((Object) subject, (Consumer) action)";
```
Comment (gayanper) [inline org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/eval/Java8Tests.java:251]: @testforstephen isn't this the same snippet you are trying in Eclipse UI which fails for you ?

Comment (testforstephen) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/eval/Java8Tests.java:251]: The snippet is same, but the evaluation context is not same. In this test case, it's stopped at Line 25. The failure case I mentioned is stopped at Line 19. It's supposed to pass as well.

**Review by testforstephen — APPROVED** [review]
LGTM

**Pull request merged**